plugins/inputs/sqlserver/connectionstring.go

```diff
@@ -0,0 +1,124 @@
+package sqlserver
+
+import (
+	"net"
+	"net/url"
+	"strings"
+)
+
+const odbcPrefix = "odbc:"
+
+const (
+	emptyServerName   = "<empty-server-name>"
+	emptyDatabaseName = "<empty-database-name>"
+)
+
+// getConnectionIdentifiers returns the sqlInstance and databaseName from the given connection string.
+// The name of the server is returned as-is in the connection string
+// If the connection string could not be parsed or sqlInstance/databaseName were not present, a placeholder value is returned
+func getConnectionIdentifiers(connectionString string) (sqlInstance string, databaseName string) {
+	if len(connectionString) == 0 {
+		return
+	}
+
+	trimmedConnectionString := strings.TrimSpace(connectionString)
+
+	if strings.HasPrefix(trimmedConnectionString, odbcPrefix) {
+		i := strings.Index(trimmedConnectionString, odbcPrefix)
+		connectionStringWithoutOdbc := trimmedConnectionString[i+len(odbcPrefix):]
+		sqlInstance, databaseName = parseConnectionStringKeyValue(connectionStringWithoutOdbc)
+	} else if strings.HasPrefix(trimmedConnectionString, "sqlserver://") {
+		sqlInstance, databaseName = parseConnectionStringURL(trimmedConnectionString)
+	} else {
+		sqlInstance, databaseName = parseConnectionStringKeyValue(trimmedConnectionString)
+	}
+
+	return
+}
+
+// parseConnectionStringKeyValue parses a "key=value;" connection string and returns the SQL instance and database name
+func parseConnectionStringKeyValue(connectionString string) (sqlInstance string, databaseName string) {
+	sqlInstance = emptyServerName
+	databaseName = emptyDatabaseName
+
+	keyValuePairs := strings.Split(connectionString, ";")
+	for _, keyValuePair := range keyValuePairs {
+		if len(keyValuePair) == 0 {
+			continue
+		}
+
+		keyAndValue := strings.SplitN(keyValuePair, "=", 2)
+		key := strings.TrimSpace(strings.ToLower(keyAndValue[0]))
+		if len(key) == 0 {
+			continue
+		}
+
+		var value string = ""
+		if len(keyAndValue) > 1 {
+			value = strings.TrimSpace(keyAndValue[1])
+		}
+
+		if isInstanceIdentifier(key) {
+			sqlInstance = value
+			if databaseName != emptyDatabaseName {
+				return
+			}
+		} else if isDatabaseIdentifier(key) {
+			databaseName = value
+			if sqlInstance != emptyServerName {
+				return
+			}
+		}
+	}
+
+	return
+}
+
+// parseConnectionStringURL parses a URL-formatted connection string and returns the SQL instance and database name
+func parseConnectionStringURL(connectionString string) (sqlInstance string, databaseName string) {
+	sqlInstance = emptyServerName
+	databaseName = emptyDatabaseName
+
+	u, err := url.Parse(connectionString)
+	if err != nil {
+		return
+	}
+
+	host, _, err := net.SplitHostPort(u.Host)
+	if err != nil {
+		// The host did not have a port
+		sqlInstance = u.Host
```

Comment:
Well there are more errors returned by `net.SplitHostPort()` e.g. indicating a malformed address etc. I think it is not save to assume that the host part didn't have a port in case of any error. So either check for the appropriate error or check for the port separator beforehand:
```suggestion
	if host, _, err := net.SplitHostPort(u.Host); err != nil {
		if ! string.Contains(u.Host, ":") {
			// The host did not have a port
			sqlInstance = u.Host
		}
```

Reply:
Looked into this more. Turns out the `Hostname()` function handles all this for me.

Reply:
Perfect solution! Please close as resolved.